workflow/engine_db.py

```diff
@@ -0,0 +1,338 @@
+
+    def __repr__(self):
+        """Allow to represent the DbWorkflowEngine."""
+        return "<DbWorkflow_engine(%s)>" % (self.name,)
+
+    def __str__(self, log=False):
+        """Allow to print the DbWorkflowEngine."""
+        return """-------------------------------
+DbWorkflowEngine
+-------------------------------
+    %s
+-------------------------------
+""" % (self.db_obj.__str__(),)
+
+    def save(self, status=None):
+        """Save the workflow instance to database."""
+        # This workflow continues a previous execution.
+        self.db_obj.save(status)
+
+    ############################################################################
+    #                                                                          #
+
+    # TODO: Modify the counters via signals instead
+    def set_counter_initial(self, obj_count):
+        """Initiate the counters of object states.
+
+        :param obj_count: Number of objects to process.
+        :type obj_count: int
+        """
+        self.db_obj.counter_initial = obj_count
+        self.db_obj.counter_halted = 0
+        self.db_obj.counter_error = 0
+        self.db_obj.counter_finished = 0
+
+    def increase_counter_halted(self):
+        """Indicate we halted the processing of one object."""
+        self.db_obj.counter_halted += 1
+
+    def increase_counter_error(self):
+        """Indicate we crashed the processing of one object."""
+        self.db_obj.counter_error += 1
+
+    def increase_counter_finished(self):
+        """Indicate we finished the processing of one object."""
+        self.db_obj.counter_finished += 1
+
+    #                                                                          #
+    ############################################################################
+
+    def abortProcessing(self):
+        """Abort current workflow execution without saving object."""
+        raise AbortProcessing
+
+    def skipToken(self):
+        """Skip current workflow object without saving it."""
+        raise SkipToken
+
+
+class DbTransitionAction(TransitionActions):
+    """Transition actions on engine exceptions for persistence object."""
+
+    @staticmethod
+    def SkipToken(obj, eng, callbacks, e):
+        """Action to take when SkipToken is raised."""
+        msg = "Skipped running this object: '%s' (object: %s)" % \
+            (str(callbacks), repr(obj))
+        eng.log.debug(msg)
+        obj.log.debug(msg)
+        raise Continue
+
+    @staticmethod
+    def AbortProcessing(obj, eng, callbacks, e):
+        """Action to take when AbortProcessing is raised."""
+        msg = "Processing was aborted: '%s' (object: %s)" % \
+            (str(callbacks), repr(obj))
+        eng.log.debug(msg)
+        obj.log.debug(msg)
+        raise Break
+
+    @staticmethod
+    def HaltProcessing(obj, eng, callbacks, e):
+        """Action to take when HaltProcessing is raised."""
+        eng.increase_counter_halted()
+        if e.action:
```

Comment:
So I am going to remove `waiting` from workflow and move it to invenio. It is now no. 4 in the enum. What I think I will do is allocate 0-9 for workflow, and 10-inf should be free for anyone to use with their database for their purposes. So I will move it to no. 10. Would it be possible to do this and add it to the database upgrade script?